Huffman.cpp

```diff
@@ -0,0 +1,426 @@
+#ifndef __PROGTEST__
+#include <cstring>
+#include <cstdlib>
+#include <cstdio>
+#include <cctype>
+#include <climits>
+#include <cassert>
+#include <cstdint>
+#include <iostream>
+#include <iomanip>
+#include <fstream>
+#include <sstream>
+#include <map>
+#include <vector>
+#include <algorithm>
+#include <set>
+#include <queue>
+#include <memory>
+#include <functional>
+#include <stdexcept>
+using namespace std;
+#define MYDEBUG
+//#define PRINTTABLE
+//#define PRINTTREE
+#endif /* __PROGTEST__ */
+
+class Tree //binary tree for storing values for huffman de/encoding
+{
+  public:
+  struct Node
+  {
+    int symbol; 
+    Node *left, 
+         *right;
+    Node() : symbol(0), left(nullptr), right(nullptr) {}
+    Node(int insert) : symbol(insert), left(nullptr), right(nullptr) {}
+    ~Node()
+    {
+      delete left;
+      delete right;
+    }
+  };
+  void print2D(const Tree::Node *root) const; //debug function to print the tree
+  private:
+  void print2DRec(const Node *root, int space) const;
+};
+
+class Huffman
+{
+  public:
+    int makeTable(ifstream* Input); //save the preorder traversal to an array
+    struct Tree::Node* makeTree(const vector<int> preorder, int n); //make a decoding tree from the array
+    void decode(ifstream* Input, ofstream* Output); //decode according to the tree
+    void initialize(ifstream* input); //prepare everything to decode
+    void printTable() const; //debug function to print table
+    //Huffman() : Root(new Tree::Node()) {}
+    ~Huffman()
+    {
+      if(initialized) delete Root;
+    }
+  private:
+    int readBit(ifstream* Input);
+    int readChar(ifstream* Input);
+    struct Tree::Node* makeTreeRec(const vector<int> preorder, int *index_ptr, int n);
+    vector<int> mTable;
+    int mBytePosition;
+    int mByte = 0;
+    bool initialized = false;
+    Tree::Node *Root;
+};
+
+//debug function to print Tree
+void Tree::print2DRec(const Tree::Node *root, int space) const
+{  
+    if (root == NULL)  
+        return;  
+    space += 3;  
+    Tree::print2DRec(root->right, space);  
+    cout<<endl;  
+    for (int i = 3; i < space; i++)  
+        cout<<" ";  
+    cout<<(char)root->symbol<<"\n";  
+    Tree::print2DRec(root->left, space);  
+}  
+  
+// Wrapper over print2DRec()  
+void Tree::print2D(const Tree::Node *root) const
```

Comment:
Can be done away with by setting `int space = 0` in header of above function (default argument value).